updater/src/updater.rs

```diff
@@ -248,189 +301,406 @@ impl Updater {
+		if state.latest.as_ref() != Some(&latest) {
+			return;
+		}
+
+		// The updated status should be set to fetching
+		if let UpdaterStatus::Fetching { ref release, binary, retries } = state.status.clone() {
+			match res {
+				// We've successfully fetched the binary
+				Ok(path) => {
+					let setup = |path: &Path| -> Result<(), String> {
+						let dest = self.updates_path(&Updater::update_file_name(&release.version));
+						if !dest.exists() {
+							info!(target: "updater", "Fetched latest version ({}) OK to {}", release.version, path.display());
+							fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| format!("Unable to create updates path: {:?}", e))?;
+							fs::copy(path, &dest).map_err(|e| format!("Unable to copy update: {:?}", e))?;
+							restrict_permissions_owner(&dest, false, true).map_err(|e| format!("Unable to update permissions: {}", e))?;
+							info!(target: "updater", "Copied updated binary to {}", dest.display());
+						}
+
+						Ok(())
 					};
 
-					info!(target: "updater", "Fetched latest version ({}) OK to {}", fetched.version, b.display());
-					fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| (format!("Unable to create updates path: {:?}", e), true))?;
-					fs::copy(&b, &dest).map_err(|e| (format!("Unable to copy update: {:?}", e), true))?;
-					restrict_permissions_owner(&dest, false, true).map_err(|e| (format!("Unable to update permissions: {}", e), true))?;
-					info!(target: "updater", "Installed updated binary to {}", dest.display());
-				}
-				let auto = match self.update_policy.filter {
-					UpdateFilter::All => true,
-					UpdateFilter::Critical if fetched.is_critical /* TODO: or is on a bad fork */ => true,
-					_ => false,
+					// There was a fatal error setting up the update, disable the updater
+					if let Err(err) = setup(&path) {
+						state.status = UpdaterStatus::Disabled;
+						warn!("{}", err);
+					} else {
+						state.status = UpdaterStatus::Ready { release: release.clone() };
+						// will lock self.state
+						drop(state);
+						self.updater_step();
+					}
+				},
+				// There was an error fetching the update, apply a backoff delay before retrying
+				Err(err) => {
+					let delay = 2usize.pow(retries) as u64;
+					// cap maximum backoff to 1 month
+					let delay = cmp::min(delay, 30 * 24 * 60 * 60);
+					let backoff = (retries, self.time_provider.now() + Duration::from_secs(delay));
+
+					state.status = UpdaterStatus::FetchBackoff { release: release.clone(), backoff, binary };
+
+					warn!("Unable to fetch update ({}): {:?}, retrying in {} seconds.", release.version, err, delay);
+				},
+			}
+		}
+	}
+
+	fn updater_step(&self) {
+		let current_block_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
+
+		let mut state = self.state.lock();
+
+		if let Some(latest) = state.latest.clone() {
```

Comment:
`if let Some(ref latest) = state.latest.clone()` should be enough, you are cloning it again anyway.

Reply:
We can't borrow since we mutate on `state`. But I've updated it to avoid cloning `latest` again when triggering the fetch. I also removed a similar `latest` clone in `poll`.